R/devmode.R

```diff
@@ -0,0 +1,251 @@
+#' Shiny Developer Mode
+#'
+#' @description \lifecycle{experimental}
+#'
+#' The methods described in this help file provide a way to change default shiny behavior when developing shiny applications (as compared to using a shiny application). For example, the following default option values are altered to have different behavior when Shiny Developer Mode is enabled:
+#'
+#' * `getOption("shiny.autoreload", default = TRUE)`: Reload the Shiny app when a sourced R file changes
+#' * `getOption("shiny.minified", default = FALSE`): Use the unminified Shiny JavaScript file, `shiny.js`
+#' * `getOption("shiny.fullstacktrace", default = TRUE`): Display the full stack trace when errors occur during Shiny app execution
+#'
+#' Other known, non-Shiny default option values that are changed when Shiny Developer Mode is enabled:
+#' * `getOption("sass.cache", default = FALSE)`: Disable `sass` caching by default. This will force `sass` to **always** compile its input values
+#'
+#'
+#' @keywords internal
+#' @describeIn devmode Function to set two options to enable/disable Shiny Developer Mode and Developer messages
+#' @param devmode Logical value which should be set to `TRUE` to enable Shiny Developer Mode
+#' @param verbose Logical value which should be set to `TRUE` display Shiny Developer messages
+#' @export
+#' @examples
+#' # Enable Shiny Developer mode
+#' devmode()
+#'
+devmode <- function(
+  devmode = getOption("shiny.devmode", TRUE),
+  verbose = getOption("shiny.devmode.verbose", TRUE)
+) {
+  options(
+    shiny.devmode = devmode,
+    shiny.devmode.verbose = verbose
+  )
+}
+
+
+#' @describeIn devmode Determines if Shiny is in Developer Mode. If the `getOption("shiny.devmode")` is set to `TRUE` and not in testing inside `testthat`, then Shiny Developer Mode is enabled.
+#' @section Avoiding direct dependency on shiny:
+#'
+#' The methods explained in this documentation file act independently from the rest of Shiny but are included to provide blue prints for your own packages. If your package already has (or willing to take) a dependency on Shiny, we recommend using the exported methods for consistent behavior.
```

Comment:
I think the first sentence should be the explanation of why `devmode()` exists. Something like, "developer mode enables a number of options to make a developer's life easier, like enabling non-minified JS and printing messages about deprecated functions and options."

We'll also need an explanation of how it searches for these values (e.g., first it looks here, then there, then ...), and why the `devmode()` function exists.